### PR TITLE
chore: Refactor command queue

### LINKED ISSUE #17446: Multidomain: Refactor command queue to be context-independent
### What would you like?

Refactor the command queue to be an independent module that does not rely on the internals or state of `cy.js`, so that it can be created in the context of a secondary domain and run.

### Why is this needed?

This is preliminary work for #16774. It's needed because currently the code assumes the command queue will be run in a single context and tied tightly to the workings of `cy.js`, but could be factored better and be a separate module. 

For the commands run in a secondary domain, the queue needs to be able to be created and run independently in the context of the secondary domain. 

The idea is to create a nice abstraction around the command queue, that is functional, stateless, and can be run regardless of context.

### Other

I separated this work from #16774 since it will be able to be merged directly into `develop` and benefit from better-factored, better-tested code.

### PR DESCRIPTION
**Pull request by chrisbreiding**
- Closes #17446

### User facing changelog

N/A - internal refactor only

### Additional Details

The general goal of this PR is to move most or all of the command queue responsibility out of `cy.js` (a huge file with a large number of responsibilities) into its own module `command_queue.js`, then to move as much as possible from that file into `queue.js`, a more functional, less stateful module that's only responsibility is being a mostly-generic queue.

This will aid with implementing #16774 (running the cypress queue in a secondary domain), since it needs to run the queue in the secondary domain just like in the primary domain, but certain things like error handling need to be forwarded and handled in the primary domain.

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue or this PR been tagged with a release in ZenHub? 
- N/A Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- N/A Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- N/A Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/src/cypress/command_queue.ts [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    // prior to ever making it through our first
+    // command
+    if (queue.stopped) {
+      return
+    }
+
+    state('current', command)
+    state('chainerId', command.get('chainerId'))
+
+    return stability.whenStable(() => {
```
Comment (flotwig) [inline packages/driver/src/cypress/command_queue.ts:139]: should we start using async/await for stuff like this? all the nested .then callbacks drive me crazy

File: packages/driver/src/cypress/command_queue.ts [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+export const create = (state, timeouts, stability, cleanup, fail, isCy) => {
+  const queue = createQueue()
+
+  const get = queue.get.bind(queue)
+  const slice = queue.slice.bind(queue)
+  const at = queue.at.bind(queue)
+  const reset = queue.reset.bind(queue)
+  const clear = queue.clear.bind(queue)
+  const stop = queue.stop.bind(queue)
```
Comment (flotwig) [inline packages/driver/src/cypress/command_queue.ts:76]: i can't help but feel that there's a more concise way to do this using classes, like `class CommandQueue extends Queue`, any reason to not do it that way?

**Review by flotwig — APPROVED** [review]
nice refactor

File: packages/driver/src/cypress/command_queue.ts (hunk 2 above)
Comment (chrisbreiding) [reply packages/driver/src/cypress/command_queue.ts]: I realized all of the binding stuff isn't necessary and was able to clean it up and make it more concise.

Comment (flotwig) [reply packages/driver/src/cypress/command_queue.ts]: well, it's not classes, but i'll take it :)

**Review by flotwig — DISMISSED** [review]

**Pull request merged**